src/ContainerTerminal.jsx

```diff
@@ -0,0 +1,187 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import cockpit from 'cockpit';
+import { Terminal } from "xterm";
+
+import * as utils from './util.js';
+import varlink from './varlink.js';
+
+import "./ContainerTerminal.css";
+
+const _ = cockpit.gettext;
+
+class ContainerTerminal extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.onChannelClose = this.onChannelClose.bind(this);
+        this.onChannelMessage = this.onChannelMessage.bind(this);
+        this.disconnectChannel = this.disconnectChannel.bind(this);
+        this.connectChannel = this.connectChannel.bind(this);
+        this.resize = this.resize.bind(this);
+
+        let term = new Terminal({
+            cols: 80,
+            rows: 24,
+            screenKeys: true,
+            cursorBlink: true,
+            fontSize: 12,
+            fontFamily: 'Menlo, Monaco, Consolas, monospace',
+            screenReaderMode: true
+        });
+
+        this.state = {
+            term: term,
+            container: props.containerId,
+            channel: null,
+            control_channel: null,
+            opened: false,
+            errorMessage: "",
+            cols: 80,
+        };
+    }
+
+    componentDidMount() {
+        this.connectChannel();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!this.state.channel && this.props.containerStatus === "running" && prevProps.containerStatus !== "running")
+            this.connectChannel();
+    }
+
+    resize(width) {
+        var padding = 11 + 5 + 50;
+        var realWidth = this.state.term._core.renderer.dimensions.actualCellWidth;
+        var cols = Math.floor((width - padding) / realWidth);
+        this.state.term.resize(cols, 24);
+        cockpit.spawn(["sh", "-c", "echo '1 24 " + cols.toString() + "'>" + this.state.control_channel], { superuser: true });
+        this.setState({ cols: cols });
+    }
+
+    componentWillUpdate(nextProps, nextState) {
```

Comment:
See https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html

This method will stop working with React 17.

Reply:
Nice catch :) It was just for resizing, I moved it to `componentDidUpdate`.